File: LeetCodeProblemsGo/LeetCode1233RemoveSubFoldersFromTheFilesystem/LeetCode1233RemoveSubFoldersFromTheFilesystem.go
package LeetCode1233RemoveSubFoldersFromTheFilesystem

func removeSubfolders(folder []string) []string {
	root := &TrieNode{
		IsFolder:   false,
		FolderName: "",
		Children:   make(map[string]*TrieNode),
	}
	for _, f := range folder {
		folderNames := getSubFolders(f)
		addFolder(root, folderNames)
	}
	result := []string{}
	dfs(root, "", &result)
	return result
}
func dfs(node *TrieNode, currnetPath string, result *[]string) {
	if len(node.FolderName) > 0 {
		currnetPath += "/" + node.FolderName
	}

	if node.IsFolder {
		*result = append(*result, currnetPath)
		return
	}
	for _, childNode := range node.Children {
		dfs(childNode, currnetPath, result)
	}
}
func getSubFolders(folder string) []string {
	result := make([]string, 0)
	buffer := ""
	length := len(folder)
	for i := 0; i < length; i++ {
		if folder[i] == '/' {
			if buffer != "" {
				result = append(result, buffer)
			}
			buffer = ""
		} else {
			buffer += string(folder[i])
		}
	}
	if buffer != "" {
		result = append(result, buffer)
	}
	return result
}
func addFolder(root *TrieNode, flodersInPath []string) {
	node := root
	for _, folder := range flodersInPath {
		if node.Children == nil {
			node.Children = make(map[string]*TrieNode)
		}
		if _, exists := node.Children[folder]; !exists {
			newNode := &TrieNode{
				IsFolder:   false,
				FolderName: folder,
				Children:   make(map[string]*TrieNode),
			}
			node.Children[folder] = newNode
			node = newNode
			continue
		}
		node = node.Children[folder]
	}
	node.IsFolder = true
}

type TrieNode struct {
	IsFolder   bool
	FolderName string
	Children   map[string]*TrieNode
}

File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode1233RemoveSubFoldersFromTheFilesystem.cpp
#include "LeetCode1233RemoveSubFoldersFromTheFilesystem.h"

#include <iostream>
#include <ostream>

#include "Auxilaries.h"
using namespace std;

vector<string> GetFolders(string& folder)
{
    vector<string> result{};
    string buffer{};
    for (const char c : folder)
    {
        if (c == '/')
        {
            if (!buffer.empty())
            {
                result.emplace_back(buffer);
                buffer.clear();
            }
            continue;
        }

        buffer += c;
    }
    if (!buffer.empty())
    {
        result.emplace_back(buffer);
        buffer.clear();
    }
    return result;
}

void AddToTrie(LeetCode1233RemoveSubFoldersFromTheFilesystem::TrieNode* root, vector<string>& folderPath)
{
    auto node = root;
    for (auto& name : folderPath)
    {
        if (node->children.contains(name))
        {
            node = node->children[name];
            continue;
        }
        auto newNode = new LeetCode1233RemoveSubFoldersFromTheFilesystem::TrieNode();
        newNode->folderName = name;
        newNode->children = {};
        node->children[name] = newNode;
        node = newNode;
    }
    node->isEndOfPath = true;
}

void AddToResult(LeetCode1233RemoveSubFoldersFromTheFilesystem::TrieNode* node, string currentPath, vector<string>& result)
{
    if (!node->folderName.empty())
        currentPath += '/' + node->folderName;
    if (node->isEndOfPath)
    {
        result.emplace_back(currentPath);
        return;
    }

    for (auto& child : node->children)
    {
        AddToResult(child.second, currentPath, result);
    }
}

vector<string> LeetCode1233RemoveSubFoldersFromTheFilesystem::removeSubfolders(vector<string>& folder)
{
    auto root = new TrieNode();
    for (string& ff : folder)
    {
        vector<string> foldersList = GetFolders(ff);
        AddToTrie(root, foldersList);
    }
    vector<string> result{};
    AddToResult(root, "", result);
    return result;
}

void LeetCode1233RemoveSubFoldersFromTheFilesystem::Run()
{
    vector<string> folder = {"/a/b/c","/a/b/ca","/a/b/d"};
    auto result = removeSubfolders(folder);
    for (auto& res : result)
    {
        cout << res << ',';
    }
    cout << endl;
}
